nbbuild/antsrc/org/netbeans/nbbuild/extlibs/CreateNoticeFile.java

```diff
@@ -0,0 +1,216 @@
+
+/**
+ * Creates an Apache compliant NOTICE file.
+ *
+ * Apache NOTICE files must always start with a preamble. This is specified in
+ * the "preamble" property, usually set to "${nb_all}/nbbuild/notice-stub.txt".
+ *
+ * The NOTICE file must include the NOTICE files of all external binaries that
+ * have one. As a consequence the NOTICE file is different for a platform build
+ * than for an ide build. When "binaryBuild" is set to true (the default) this
+ * task will scan all the appropriate modules in the build and automatically
+ * include the "[module]external/*notice.txt" and "[module]/*notice.txt" files,
+ * if they exist.
+ *
+ * If "binaryBuild" is not set to true then only the preamble and the
+ * "[module]/*notice.txt" files will be included in the final NOTICE file.
+ *
+ */
+public class CreateNoticeFile extends Task {
+
+    private boolean verbose = false;
+
+    public void setVerbose(boolean verbose) {
+        this.verbose = verbose;
+    }
+
+    private boolean binaryBuild = true;
+
+    public void setBinaryBuild(boolean binaryBuild) {
+        this.binaryBuild = binaryBuild;
+    }
+
+    private File preamble;
+
+    public void setPreamble(File preamble) {
+        this.preamble = preamble;
+    }
+
+    private ArrayList<FileSet> fileSets = new ArrayList<FileSet>();
+
+    public void addFileSet(FileSet fs) {
+        fileSets.add(fs);
+    }
+
+    private File destFile;
+
+    public void setDestFile(File destFile) {
+        this.destFile = destFile;
+    }
+
+    private static final void append(PrintWriter output, File file) throws IOException {
+        /* Appends a file to output, rejecting multiple consecutive empty lines. */
+        boolean previousLineEmpty = true;
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file), "utf-8"))) {
+            do {
+                String line = reader.readLine();
+                if (line == null) {
+                    break;
+                }
+                line = line.trim();
+                boolean empty = line.length() == 0;
+                if (empty && previousLineEmpty) {
+                    // Skip line
+                } else {
+                    previousLineEmpty = empty;
+                    output.println(line);
+                }
+            } while (true);
+        }
+    }
+
+    @Override
+    public void execute() throws BuildException {
+
+        if (getProject().getProperty("nb_all") == null) {
+            throw new BuildException("Please specify a nb_all directory.");
+        }
+
+        try {
+            PrintWriter output = new PrintWriter(new OutputStreamWriter(new FileOutputStream(destFile), "utf-8"));
+
+            append(output, preamble);
+
+            Hashtable<String, Object> properties = getProject().getProperties();
+
+            // This excerpt from ModuleListsParser.java
+            Set<String> standardModules = new TreeSet<String>();
```

Comment:
In the nbbuild/build.xml, there's a property "allmodules" that (AFAIK) contains the modules in the selected config. Would be better to just send the value of the property to the task rather than trying to recompute the value again.

Reply:
Definitely. Thanks for that. The "allmodules" name was confusing me: I thought that were all the modules, not all the modules in the current configuration.